target/linux/realtek/files-6.12/drivers/i2c/busses/i2c-rtl9300.c

```diff
@@ -141,20 +141,25 @@ static int i2c_read(void __iomem *r0, u8 *buf, int len)
 
 static int i2c_write(void __iomem *r0, u8 *buf, int len)
 {
+	u32 vals[4] = {};
+	int i;
+
 	if (len > 16)
 		return -EIO;
 
-	for (int i = 0; i < len; i++) {
-		u32 v;
+	for (i = 0; i < len; i++) {
+		unsigned int shift = (i % 4) * 8;
+		unsigned int reg = i / 4;
 
-		if (! (i % 4))
-			v = 0;
-		v <<= 8;
-		v |= buf[i];
-		if (i % 4 == 3 || i == len - 1)
-			writel(v, r0 + (i / 4) * 4);
+		if (i % 4 == 0)
+			vals[reg] = 0;
```

Comment:
Before somebody asks, this is also how the upstream driver does this - don't ask me why. The `u32 vals[4] = {};` already forces a 0 initialization of the `vals` array on the stack. When the `= {}` would be missing, the array would not be initialized.

Reply:
Sorry, this bothered me too much. I've resent the upstream patch series and adjusted the OpenWrt code to drop this.